src/threshold_sign.rs

```diff
@@ -99,26 +101,42 @@ impl<N: NodeIdT> DistAlgorithm for ThresholdSign<N> {
 
 impl<N: NodeIdT> ThresholdSign<N> {
     /// Creates a new instance of `ThresholdSign`, with the goal to collaboratively sign `msg`.
-    pub fn new<M: AsRef<[u8]>>(netinfo: Arc<NetworkInfo<N>>, msg: M) -> Self {
+    pub fn new(netinfo: Arc<NetworkInfo<N>>) -> Self {
         ThresholdSign {
             netinfo,
-            msg_hash: hash_g2(msg),
+            msg_hash: None,
             received_shares: BTreeMap::new(),
             had_input: false,
             terminated: false,
         }
     }
 
+    /// Sets msg_hash and resets internal state accordingly.
+    pub fn set_message<M: AsRef<[u8]>>(&mut self, msg: M) -> Result<()> {
+        if self.msg_hash.is_some() {
+            return Err(Error::MultipleMessagesToSign);
+        }
+        self.msg_hash = Some(hash_g2(msg));
+        Ok(())
+    }
+
     /// Sends our signature shares, and if we have collected enough, returns the full signature.
+    /// Returns an empty step if the message to sign hasn't been received yet.
     pub fn sign(&mut self) -> Result<Step<N>> {
-        if self.had_input {
+        if self.had_input || self.msg_hash.is_none() {
             return Ok(Step::default());
         }
         self.had_input = true;
+        let mut step = Step::default();
+        step.fault_log.extend(self.remove_invalid_shares());
         if !self.netinfo.is_validator() {
             return self.try_output();
         }
-        let msg = Message(self.netinfo.secret_key_share().sign_g2(self.msg_hash));
+        let msg = Message(
+            self.netinfo
+                .secret_key_share()
+                .sign_g2(self.msg_hash.unwrap()),
```

Comment:
Let's take the hash above, at the same point where we're returning an error if it's `None`, to avoid the `unwrap`.